TooManyClauses should be 4xx

```
[es-es-search-76956fd674-vfnbq][100.65.76.186:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.search.query.QueryPhaseExecutionException: Query Failed [Failed to execute main query]
\tat org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:229)
\tat org.elasticsearch.search.query.QueryPhase.executeQuery(QueryPhase.java:135)
\tat org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:62)
\tat org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:532)
\tat org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:690)
\tat org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:559)
\tat org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:78)
\tat org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:75)
\tat org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:100)
\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
\tat org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:991)
\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
\tat java.lang.Thread.run(Thread.java:1570)
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: too_many_nested_clauses: Query contains too many nested clauses; maxClauseCount is set to 32768
\tat org.apache.lucene.search.IndexSearcher$1.consumeTerms(IndexSearcher.java:822)
\tat org.apache.lucene.queries.spans.SpanTermQuery.visit(SpanTermQuery.java:99)
\tat org.apache.lucene.queries.spans.SpanNearQuery.visit(SpanNearQuery.java:272)
\tat org.apache.lucene.queries.spans.SpanOrQuery.visit(SpanOrQuery.java:91)
\tat org.apache.lucene.search.BooleanQuery.visit(BooleanQuery.java:616)
\tat org.apache.lucene.search.IndexSearcher.rewrite(IndexSearcher.java:782)
\tat org.elasticsearch.search.internal.ContextIndexSearcher.rewrite(ContextIndexSearcher.java:200)
\tat org.elasticsearch.search.internal.SearchContext.rewrittenQuery(SearchContext.java:277)
\tat org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:157)
```

--- a/docs/changelog/112678.yaml
+++ b/docs/changelog/112678.yaml
@@ -1,0 +1,6 @@
+pr: 112678
+summary: Make "too many clauses" throw IllegalArgumentException to avoid 500s
+area: Search
+type: bug
+issues:
+ - 112177

--- a/server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
+++ b/server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java
@@ -202,6 +202,8 @@ public class ContextIndexSearcher extends IndexSearcher implements Releasable {
         } catch (TimeExceededException e) {
             timeExceeded = true;
             return new MatchNoDocsQuery("rewrite timed out");
+        } catch (TooManyClauses e) {
+            throw new IllegalArgumentException("Query rewrite failed: too many clauses", e);
         } finally {
             if (profiler != null) {
                 profiler.stopAndAddRewriteTime(rewriteTimer);

--- a/server/src/test/java/org/elasticsearch/search/query/QueryPhaseTests.java
+++ b/server/src/test/java/org/elasticsearch/search/query/QueryPhaseTests.java
@@ -29,6 +29,7 @@ import org.apache.lucene.index.NoMergePolicy;
 import org.apache.lucene.index.Term;
 import org.apache.lucene.queries.spans.SpanNearQuery;
 import org.apache.lucene.queries.spans.SpanTermQuery;
+import org.apache.lucene.search.BooleanClause;
 import org.apache.lucene.search.BooleanClause.Occur;
 import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.Collector;
@@ -1103,6 +1104,22 @@ public class QueryPhaseTests extends IndexShardTestCase {
             NEVER_CACHE_POLICY,
             true
         );
+    }
+
+    public void testTooManyClauses() throws Exception {
+        indexDocs();
+        var oldCount = IndexSearcher.getMaxClauseCount();
+        try {
+            var query = new BooleanQuery.Builder().add(new BooleanClause(new MatchAllDocsQuery(), Occur.SHOULD))
+                .add(new MatchAllDocsQuery(), Occur.SHOULD)
+                .build();
+            try (TestSearchContext context = createContext(newContextSearcher(reader), query)) {
+                IndexSearcher.setMaxClauseCount(1);
+                expectThrows(IllegalArgumentException.class, context::rewrittenQuery);
+            }
+        } finally {
+            IndexSearcher.setMaxClauseCount(oldCount);
+        }
     }
 
     private static ContextIndexSearcher noCollectionContextSearcher(IndexReader reader) throws IOException {

